pkg/cluster/hash.go

```diff
@@ -10,51 +10,81 @@ import (
+	"github.com/sirupsen/logrus"
+
 	"github.com/openshift/openshift-azure/pkg/api"
 	"github.com/openshift/openshift-azure/pkg/arm"
+	"github.com/openshift/openshift-azure/pkg/config"
+	"github.com/openshift/openshift-azure/pkg/util/writers"
 )
 
 type Hasher interface {
-	HashScaleSet(*api.OpenShiftManagedCluster, *api.AgentPoolProfile) ([]byte, error)
+	HashWorkerScaleSet(*api.OpenShiftManagedCluster, *api.AgentPoolProfile) ([]byte, error)
+	HashMasterScaleSet(cs *api.OpenShiftManagedCluster, app *api.AgentPoolProfile, instanceID int64) ([]byte, error)
 }
 
 type hasher struct {
+	log        *logrus.Entry
 	testConfig api.TestConfig
 }
 
-// hashScaleSets returns the set of desired state scale set hashes
-func (h *hasher) HashScaleSet(cs *api.OpenShiftManagedCluster, app *api.AgentPoolProfile) ([]byte, error) {
+func (h *hasher) vmss(cs *api.OpenShiftManagedCluster, app *api.AgentPoolProfile) (*compute.VirtualMachineScaleSet, error) {
 	// the hash is invariant of name, suffix, count
 	appCopy := *app
 	appCopy.Count = 0
 	appCopy.Name = ""
 
-	vmss, err := arm.Vmss(cs, &appCopy, "", "", h.testConfig) // TODO: backupBlob is rather a layering violation here
+	return arm.Vmss(cs, &appCopy, "", "", h.testConfig) // TODO: backupBlob is rather a layering violation here
+}
+
+// HashWorkerScaleSet returns the hash of a worker scale set
+func (h *hasher) HashWorkerScaleSet(cs *api.OpenShiftManagedCluster, app *api.AgentPoolProfile) ([]byte, error) {
+	hash := sha256.New()
+
+	vmss, err := h.vmss(cs, app)
 	if err != nil {
 		return nil, err
 	}
 
-	data, err := json.Marshal(vmss)
+	err = json.NewEncoder(hash).Encode(vmss)
 	if err != nil {
 		return nil, err
 	}
 
-	if app.Role == api.AgentPoolProfileRoleMaster {
-		// add certificates pulled from keyvault by the master to the hash, to
-		// ensure the masters update if a cert changes.  We don't add the keys
-		// because these are not necessarily stable (sometimes the 'D' value of
-		// the RSA key returned by keyvault differs to the one that was sent).
-		// I believe that in a given RSA key, there are multiple suitable values
-		// of 'D', so this is not a problem, however it doesn't make the value
-		// suitable for a hash.  References:
-		// https://stackoverflow.com/a/14233140,
-		// https://crypto.stackexchange.com/a/46572.
-		data = append(data, cs.Config.Certificates.OpenShiftConsole.Certs[0].Raw...)
-		data = append(data, cs.Config.Certificates.Router.Certs[0].Raw...)
+	return hash.Sum(nil), nil
+}
+
+// HashMasterScaleSet returns the hash of a worker scale set
+func (h *hasher) HashMasterScaleSet(cs *api.OpenShiftManagedCluster, app *api.AgentPoolProfile, instanceID int64) ([]byte, error) {
+	hash := sha256.New()
+
+	vmss, err := h.vmss(cs, app)
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.NewEncoder(hash).Encode(vmss)
+	if err != nil {
+		return nil, err
+	}
+
+	err = arm.WriteStartupFiles(h.log, cs, writers.NewTarWriter(hash), config.GetHostname(app, "", instanceID), "")
```

Comment:
I imagine this is using same writer to append data to it?
Was this main motivation to move to writer based function?

Reply:
yes, all the writer work is leading up to this line :)